source/imaer-shared/src/main/java/nl/overheid/aerius/shared/emissions/FarmlandEmissionsCalculator.java

```diff
@@ -35,38 +37,37 @@ public FarmlandEmissionsCalculator(final FarmlandEmissionFactorSupplier farmland
 
   public Map<Substance, Double> updateEmissions(final FarmlandEmissionSource farmlandSource) {
     final Map<Substance, BigDecimal> summed = new EnumMap<>(Substance.class);
-    for (final FarmlandActivity activity : farmlandSource.getSubSources()) {
-
-      if (activity instanceof FarmlandGrazingActivity) {
-        updateEmissions((FarmlandGrazingActivity) activity, summed);
-      } else {
-        updateEmissions(activity.getEmissions(), summed);
-      }
+    for (final AbstractFarmlandActivity activity : farmlandSource.getSubSources()) {
+      activity.accept(this, summed);
     }
 
     final Map<Substance, Double> result = new EnumMap<>(Substance.class);
     summed.forEach((key, value) -> result.put(key, value.doubleValue()));
     return result;
   }
 
-  private void updateEmissions(final Map<Substance, Double> emissions, final Map<Substance, BigDecimal> summedEmissions) {
-    emissions.forEach(
+  @Override
+  public void visit(final CustomFarmlandActivity activity, final Map<Substance, BigDecimal> summedEmissions) {
+    activity.getEmissions().forEach(
         (key, value) -> summedEmissions.merge(key, BigDecimal.valueOf(value), (v1, v2) -> v1.add(v2)));
   }
 
-  private void updateEmissions(final FarmlandGrazingActivity activity, final Map<Substance, BigDecimal> summedEmissions) {
-    Map<Substance, Double> repositoryEmissions = farmlandEmissionFactorSupplier.getGrazingEmissionFactors(activity.getGrazingCategoryCode());
-    Map<Substance, Double> userEmissions = activity.getEmissions();
-
-    // Prefer emission factors supplied by the farmlandEmissionFactorSupplier
-    Map<Substance, Double> activityEmissionFactors = repositoryEmissions.isEmpty() ? userEmissions : repositoryEmissions;
+  @Override
+  public void visit(final StandardFarmlandActivity activity, final Map<Substance, BigDecimal> summedEmissions) {
+    final Map<Substance, Double> emissionFactors = farmlandEmissionFactorSupplier
+        .getFarmSourceEmissionFactors(activity.getFarmSourceCategoryCode());
 
     if (activity.getFarmEmissionFactorType() == FarmEmissionFactorType.PER_ANIMAL_PER_DAY) {
-      activityEmissionFactors.forEach((key, value) ->
-          summedEmissions.merge(key, BigDecimal.valueOf(value).multiply(BigDecimal.valueOf(activity.getDays())), (v1, v2) -> v1.add(v2))
+      emissionFactors.forEach((key, value) ->
+          summedEmissions.merge(key, BigDecimal.valueOf(value).multiply(BigDecimal.valueOf(activity.getNumberOfDays()))
+              .multiply(BigDecimal.valueOf(activity.getNumberOfAnimals())), (v1, v2) -> v1.add(v2))
+      );
+    }
+
+    if (activity.getFarmEmissionFactorType() == FarmEmissionFactorType.PER_ANIMAL_PER_YEAR) {
+      emissionFactors.forEach((key, value) ->
+          summedEmissions.merge(key, BigDecimal.valueOf(value).multiply(BigDecimal.valueOf(activity.getNumberOfAnimals())), (v1, v2) -> v1.add(v2))
       );
-    } else {
-      updateEmissions(activityEmissionFactors, summedEmissions);
```

Comment:
might be an idea to use `else if` and a final `else` clause where you throw an exception if the type is unknown?